[mlir][llvm] Use zeroinitializer for TargetExtType

sommerlukas opened the pull request:
Use the recently introduced llvm.mlir.zero operation for values with LLVM target extension type. Replaces the previous workaround that uses a single zero-valued integer attribute constant operation.

sommerlukas committed:
[mlir][llvm] Use zeroinitializer for TargetExtType

Use the recently introduced llvm.mlir.zero operation for values with
LLVM target extension type. Replaces the previous workaround that uses a
single zero-valued integer attribute constant operation.

Signed-off-by: Lukas Sommer <lukas.sommer@codeplay.com>

zero9178 commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -2051,13 +2051,8 @@ LogicalResult GlobalOp::verify() {
-      // zeroinitializer operation in the LLVM dialect.
-      if (!isa<IntegerAttr>(value) || !isZeroAttribute(value))
-        return emitOpError()
-               << "expected zero value for global with target extension type";
+      return emitOpError() << "global with target extension type can only be "
+                              "initialized with zero-initializer";
     }
```
ultra nit: drop the braces here

sommerlukas replied:
Done.

zero9178 commented:
Very happy to see this being cleaned up!

I was initially surprised that this PR did not require any changes in the export and checked the original revision again. It added code specifically for initializing target extension types here: 
https://github.com/llvm/llvm-project/blob/main/mlir/lib/Target/LLVMIR/ModuleTranslation.cpp#L347-L356
Can this be removed now?

I see that the export logic is already implemented in `llvm.mlir.zero` op.

sommerlukas commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -2510,6 +2505,21 @@ Region *LLVMFuncOp::getCallableRegion() {
+             << "target extension type does not support zero-initializer";
+
+    return success();
+  }
```
Done.

sommerlukas commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -2554,16 +2564,7 @@ LogicalResult LLVM::ConstantOp::verify() {
-             << "only zero-initializer allowed for target extension types";
-
-    return success();
+    return emitOpError() << "target extension type does not support constants.";
```
I changed the error message as you proposed. 

Using the fall-through wouldn't work without modification, because we wouldn't be catching things like this anymore:
```
%0 = llvm.mlir.constant(0 : index) : !llvm.target<"spirv.Event">
```

Modifying the if below would allow to fall-though, but I think the error message in that case does not capture the intention:
```
"only supports integer, float, string or elements attributes"
```
The attribute itself isn't the problem, it's more that constant operations do not support the type at all.

Therefore, I've kept the separate if for the target extension type and only changed the error message as proposed.

sommerlukas commented:
> Very happy to see this being cleaned up!

Thanks for your feedback @zero9178!

> I was initially surprised that this PR did not require any changes in the export and checked the original revision again. It added code specifically for initializing target extension types here: https://github.com/llvm/llvm-project/blob/main/mlir/lib/Target/LLVMIR/ModuleTranslation.cpp#L347-L356 Can this be removed now?
> 
> I see that the export logic is already implemented in `llvm.mlir.zero` op.

Good catch, I removed that from `ModuleTranslation.cpp`.

zero9178 reviewed (approved):
LGTM!

sommerlukas commented:
Rebased on `main`, squashed and ran tests (`check-mlir`). 

@zero9178 Can you please commit/merge this PR for me, I do not (yet) have commit access?
